src/ORM/Related/DataObjectRelatedService.php

```diff
@@ -0,0 +1,332 @@
+                if (empty($configData)) {
+                    continue;
+                }
+                foreach ($configData as $component => $componentClass) {
+                    // Example $component: 'MyFile'
+                    // Example $componentClassName: SilverStripe\Assets\File::class
+                    // Example php file config: private static $has_one = [ 'MyFile' => File::class ]
+                    $componentClass = $this->updateComponentClass($componentClass, $config, $throughClasses);
+
+                    // $record with usually be a File, or a subclass of File. Example: MyFile
+                    // $componentClass will usually be a File, so a $record subclass like MyFile will match
+                    if (!($record instanceof $componentClass)) {
+                        continue;
+                    }
+
+                    // Query database to get file usage data
+                    $results = [];
+                    if ($relationType === 'has_one') {
+                        // Ensure table exists, this is required for TestOnly SapphireTest classes
+                        if (!isset($classToTableName[$class])) {
+                            continue;
+                        }
+
+                        // Example: class MyPage ... private static $has_one = [ 'MyFile' => File::class ]
+                        $componentIDField = "{$component}ID";
+
+                        // Only get database fields from the current class model, not parent class model
+                        $dbFields = $dataObjectSchema->databaseFields($class, false);
+                        if (!isset($dbFields[$componentIDField])) {
+                            continue;
+                        }
+                        $tableName = $dataObjectSchema->tableName($class);
+
+                        // Example SQL: SELECT "ID" FROM "MyPage" WHERE "MyFileID" = 123;
+                        $results = SQLSelect::create(
+                            '"ID"',
+                            sprintf('"%s"', $tableName),
+                            sprintf('"%s" = %s', $componentIDField, $record->ID)
+                        )->execute();
+                    }
+                    if ($relationType === 'many_many' || $relationType === 'belongs_many_many') {
+                        $results = $this->fetchManyManyResults(
+                            $record,
+                            $class,
+                            $component,
+                            false,
+                            $dataObjectSchema,
+                            $classToTableName
+                        );
+                    }
+                    $this->updateClassIDs($classIDs, $results, $class);
+                }
+            }
+        }
+
+        // Loop has_one relationships on the DataObject we're getting usage for
+        // e.g. File.has_one = Page, Page.has_many = File
+        $class = get_class($record);
```

Comment:
Your record already has all the methods needed to retrieve relation data. Reading it from the config requires the duplication of logic.

Reply:
Updated